godtossing/Pos/Ssc/GodTossing/Workers.hs

```diff
@@ -221,43 +226,67 @@ onNewSlotCommitment slotId@SlotId {..} sendActions
 -- Openings-related part of new slot processing
 onNewSlotOpening
     :: (GtMessageConstraints, SscMode SscGodTossing ctx m)
-    => SlotId -> Worker m
-onNewSlotOpening SlotId {..} sendActions
+    => GtOpeningParams -> SlotId -> Worker m
+onNewSlotOpening params SlotId {..} sendActions
     | not $ isOpeningIdx siSlot = pass
     | otherwise = do
         ourId <- getOurStakeholderId
         globalData <- gtGetGlobalState
         unless (hasOpening ourId globalData) $
             case globalData ^. gsCommitments . to getCommitmentsMap . at ourId of
                 Nothing -> logDebug noCommMsg
-                Just _  -> onNewSlotOpeningDo ourId
+                Just _  -> SS.getOurOpening siEpoch >>= \case
+                    Nothing   -> logWarning noOpenMsg
+                    Just open -> sendOpening ourId open
   where
     noCommMsg =
-        "We're not sending opening, because there is no commitment from us in global state"
-    onNewSlotOpeningDo ourId = do
-        mbOpen <- SS.getOurOpening siEpoch
-        case mbOpen of
-            Just open -> do
-                let msg = MCOpening ourId open
-                sscProcessOurMessage (sscProcessOpening ourId open)
-                sendOurData (enqueueMsg sendActions) OpeningMsg ourId msg siEpoch 2
-            Nothing -> logWarning "We don't know our opening, maybe we started recently"
+        "We're not sending opening, because there is no commitment \
+        \from us in global state"
+    noOpenMsg =
+        "We don't know our opening, maybe we started recently"
+    sendOpening ourId open = do
+        mbOpen' <- case params of
+            GtOpeningNone   -> pure Nothing
+            GtOpeningNormal -> pure (Just open)
+            GtOpeningWrong  -> do
+                keys <- NE.fromList . map (asBinary . toVssPublicKey) <$>
```

Comment:
Rather complicated... Why not `arbitrary` opening?

Reply:
Okay, changed.